[ADAM-1635] Eliminate passing FASTQ splittable status via config.

fnothaft opened the pull request:
Resolves #1635. Instead of passing whether a FASTQ was splittable via config, checks to see if the compression codec is splittable. This is more reliable. In the case of a .gz file, the BGZFEnhancedGZipCodec properly handles this edge case by checking the stream type; this coupled with us explicitly checking the stream when split picking ensures that we don't try to create an invalid GZIP split. Additionally, I identified and fixed an error in the old FASTQ code that did a seek on the uncompressed input stream to backtrack if seeing a line of quality scores that began with @ when identifying the position of the first valid record in a split. Instead, we check for two successive lines that start with an @, which indicates that the first line contains quality scores, while the second line contains read names.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/2281/
Test PASSed.

heuermh commented on adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java:
```diff
@@ -145,12 +145,23 @@ protected FastqRecordReader(final Configuration conf,
-        // a contract where it will put the file's splittable status into the hadoop
-        // configuration object.
-        isSplittable = conf.getBoolean(FastqInputFormat.FILE_SPLITTABLE, false);
+	// if our codec is splittable, we can (tentatively) say that
```
Yer editor accidentally used tabs for some of these lines

fnothaft replied:
Thanks for catching this; I was editing this patch on a different computer from my usual and I was wondering why the diff looked weird.

heuermh commented on adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java:
```diff
@@ -206,20 +217,19 @@ protected final int positionAtFirstRecord(final FSDataInputStream stream,
             // see above note about 
             // SplittableCompressionCodec.createInputStream needing the stream
             // to be at offset 0
-            stream.seek(0);
```
the comment above this line can be removed

fnothaft replied:
I think this is still useful info to keep around, but I'll update the comment to better reflect the changed code.

fnothaft commented:
Pushed a commit addressing reviewer comments.

BTW @heuermh do you think it would be worthwhile to add something to our CI that would flag any tabs in our source and fail the build? I would've missed those if you hadn't caught them.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/2284/
Test PASSed.

heuermh commented:
> do you think it would be worthwhile to add something to our CI that would flag any tabs in our source and fail the build? I would've missed those if you hadn't caught them.

We have a linter that runs on the scala source, this made it through because it was a java source file.  I don't think we can put a CI check on the whole repo because some of our test resources require tab characters.

heuermh commented:
Thank you, @fnothaft

heuermh commented:
Sorry, wrong button, I should've squashed.

fnothaft commented:
> We have a linter that runs on the scala source, this made it through because it was a java source file. I don't think we can put a CI check on the whole repo because some of our test resources require tab characters.

I mean, sure, but we could do something like:

```
find adam-*/src -name "*.java" -exec ./scripts/failIfHasTabs.sh {} \;
find adam-*/src -name "*.R" -exec ./scripts/failIfHasTabs.sh {} \;
find adam-*/src -name "*.py" -exec ./scripts/failIfHasTabs.sh {} \;
```

heuermh commented:
+1, add `*.pom`, `*.sh`
